File: app/src/main/java/com/adlab/balda/model/OneManGame.java
package com.adlab.balda.model;

import com.adlab.balda.database.WordsDataSource;
import com.adlab.balda.enums.FieldSizeType;
import com.adlab.balda.enums.FieldType;
import com.adlab.balda.model.field.AbstractField;
import com.adlab.balda.model.field.ClassicField;
import com.adlab.balda.model.field.HexagonField;

import java.util.ArrayList;
import java.util.List;

public class OneManGame {

    private static final char EMPTY_CELL_VALUE = ' ';

    //private char[] field;
    //private FieldType fieldType;
    private AbstractField field;
    private int rowCount;
    private int colCount;
    //private String initWord;
    private GamePlayer player;

    private WordsDataSource wordsRepository;

    public OneManGame(String initWord, FieldSizeType fieldSize, FieldType fieldType, GamePlayer player, WordsDataSource wordsRepository) {
        this.rowCount = fieldSize.intValue();
        this.colCount = fieldSize.intValue();
        //this.initWord = initWord;
        //this.fieldType = fieldType;
        this.player = player;
        this.wordsRepository = wordsRepository;
        if (fieldType == FieldType.SQUARE) {
            field = new ClassicField(initWord, fieldSize);
        } else {
            field = new HexagonField(initWord, fieldSize);
        }
    }

    public int getRowCount() {
        return rowCount;
    }

    public int getColCount() {
        return colCount;
    }

    public char[] getCopyOfCells() {
        return field.getCopyOfCells();
    }

    public GamePlayer getPlayer() {
        return player;
    }

    public List<String> getUsedWords() {
        ArrayList<String> list = new ArrayList<>();
        list.add(field.getInitWord());
        list.addAll(player.getEnteredWords());
        return list;
    }

    public void makeMove(final int enteredCellNumber, final char enteredLetter, int[] cellsWithWord, final MakeMoveCallback callback) {
        final String enteredWord = field.getEnteredWord(enteredCellNumber, enteredLetter, cellsWithWord);

        if (field.getInitWord().equals(enteredWord)) {
            callback.onWordIsAlreadyUsed();
            return;
        }

        for (String word : player.getEnteredWords()) {
            if (word.equals(enteredWord)) {
                callback.onWordIsAlreadyUsed();
                return;
            }
        }

        wordsRepository.isWordExist(enteredWord, new WordsDataSource.CheckWordCallback() {
            @Override
            public void onWordChecked(boolean isWordExist) {
                if (isWordExist) {
                    boolean isGameFinished = applyMove(enteredCellNumber, enteredLetter, enteredWord);
                    if (isGameFinished) {
                        callback.onGameFinished();
                    } else {
                        callback.makeNextMove();
                    }
                } else {
                    callback.onWordIsNotExist();
                }
            }
        });
    }

    // return true if game is finished and false otherwise
    private boolean applyMove(int enteredCellNumber, char enteredLetter, String enteredWord) {
        player.increaseScore(enteredWord.length());
        player.addEnteredWord(enteredWord);
        field.setLetter(enteredCellNumber, enteredLetter);
        return isGameFinish();
    }

    private boolean isGameFinish() {
        return field.isFieldFull();
    }

    public FieldType getFieldType() {
        return field.getFieldType();
    }

    public FieldSizeType getFieldSize() {
        return field.getFieldSize();
    }

    public interface MakeMoveCallback {
        void makeNextMove();
        void onWordIsNotExist();
        void onWordIsAlreadyUsed();
        void onGameFinished();
    }
}

File: app/src/main/java/com/adlab/balda/model/OneManGame.kt
package com.adlab.balda.model

import com.adlab.balda.database.AppRepository
import com.adlab.balda.model.field.AbstractField
import java.util.*

class OneManGame(
        val field: AbstractField,
        val player: GamePlayer,
        private val repository: AppRepository
) {
    val copyOfCells
            get() = field.getCopyOfCells()

    val usedWords
        get() = ArrayList<String>().apply {
            this.add(field.initWord)
            this.addAll(player.enteredWords)
        }

    suspend fun makeMove(enteredCellNumber: Int, enteredLetter: Char, cellsWithWord: IntArray,
                         callback: MakeMoveCallback) {
        val enteredWord = field.getEnteredWord(enteredCellNumber, enteredLetter, cellsWithWord)

        if (field.initWord == enteredWord) {
            callback.onWordIsAlreadyUsed()
            return
        }
        player.enteredWords.forEach { word ->
            if (word == enteredWord) {
                callback.onWordIsAlreadyUsed()
                return
            }
        }
        if (repository.isWordExist(enteredWord)) {
            val isGameFinished = applyMove(enteredCellNumber, enteredLetter, enteredWord)
            if (isGameFinished) {
                callback.onGameFinished()
            } else {
                callback.makeNextMove()
            }
        } else {
            callback.onWordIsNotExist();
        }
    }

    // return true if game is finished and false otherwise
    private fun applyMove(enteredCellNumber: Int, enteredLetter: Char, enteredWord: String): Boolean {
        player.addEnteredWord(enteredWord)
        field.setLetter(enteredCellNumber, enteredLetter)
        return isGameFinished
    }

    private val isGameFinished
        get() = field.isFieldFull

    val fieldType
        get() = field.fieldType

    val fieldSize
        get() = field.fieldSize

    interface MakeMoveCallback {
        fun makeNextMove()
        fun onWordIsNotExist()
        fun onWordIsAlreadyUsed()
        fun onGameFinished()
    }
}